### PR TITLE
brew update: show substantially changed formulae only

### PR DESCRIPTION
**Pull request by bfontaine**
This is a follow-up to @mikemcquaid’s [comment here](https://github.com/Homebrew/homebrew/pull/35458#issuecomment-68517277).

The goal is to only report substantially changed formulae when running `brew update`, that is, not include formulae with only minor coding style changes. This PR introduces an ~~`Homebrew.substantial_change` method which tests if a given path has been *substantially* changed since the last update, called by `Homebrew.report` to determine if a formula should be included in the report.~~ `Homebrew.updated_paths` method which returns a set of updated paths since the last update. It works by parsing the output of `git log` and excluding commits which include `#homebrew-update` in their message (subject or body).

Some notes:
* ~~`substantial_change` works with a patterns blacklist. I originally used a whitelist but it was too restrictive and false negatives (including formulae with minor changes) are better than false positives (excluding formulae with substantial changes)~~
* ~~it only tests the last commit for each formula. I tested with more commits but it really slow down the update. The downside of this is that a formula with two commits, a version bump *then* a minor change, will be rejected. There may be a better/faster way to do that than what I’m using (`git log rev1 rev2 path`)~~
* I don’t know how to test this feature. `test_updater.rb` generates diffs for tests but I don’t know how to mock commit messages.

### CODE REVIEW

File: Library/Homebrew/cmd/update.rb [hunk 1]
```diff
@@ -218,6 +220,24 @@ def diff
...
+      "git", "log", "--pretty=format:%s", "-n", max.to_s,
+      initial_revision, current_revision, path
+    ).split(/\n/)
+  end
+
+  # Test if a formula file has been substantially changed, e.g. version bump or
+  # updated bottle.
+  def substantial_change(path)
+    unacceptable_patterns = [
+      /: modernize\.?$/,
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/update.rb]: Maybe add `add test`

**Comment by xu-cheng** [comment]
Instead of using regex, using version/reversion/url/sha1 changing may be a better approach. That is, similar to what `brew upgrade` will pick up.

**Comment by mikemcquaid** [comment]
I think a better approach may be to have a keyword `#homebrew-ignore` or something that we add to commit messages for stuff to be ignored.

**Comment by xu-cheng** [comment]
:+1 for @mikemcquaid's idea.

**Comment by bfontaine** [comment]
@mikemcquaid I just edited for that, and added a faster way to check commits (it executes `git log` only once instead of once per formula). It now breaks tests because `brew update` only includes files with commits, and from what I see your tests use mock diffs but not mock commits (I may be wrong here).

The first two notes in my initial comment aren’t true anymore, the last one is still true.

**Comment by mikemcquaid** [comment]
Interested in thoughts from other @Homebrew/owners here.

**Comment by jacknagel** [comment]
I don't understand the upside. I dislike style-only commits as a rule, not because they clutter the update output.

**Comment by mikemcquaid** [comment]
@jacknagel I think the advantage of an approach like is allowing us to do e.g. automated rubocop style improvements across many formulae at once (like hundreds or thousands) without making the `brew update` status useless.

**Comment by jacknagel** [comment]
I firmly believe "blame-ability" is much more important that changing single quotes to double quotes, and would strongly oppose such mass style changes. But even if we were going to do so, I think the approach is wrong. I think it would make more sense to drop the "updated formula" output entirely, and replace it with something like #25951.

**Comment by mikemcquaid** [comment]
@jacknagel Yeh, that makes sense to me. We can discuss mass style changes somewhere else but I think getting #25951 merged with your comments would be a better thing.

@bfontaine want to give that a go? Sorry to keep changing things here and feel free to say no or ask questions. Thanks again for your work!

**Comment by bfontaine** [comment]
@mikemcquaid in the end it’s your repo, you decide what you want in it, but I agree with @jacknagel regarding the importance of blame-ability and replace the “updated formula” output by #25951.

I’d be happy to try it, but wouldn’t you prefer than @ranmocy edit their pull-request instead of me basically re-coding his idea?

Btw, do you have a list somewhere of things we, non-core-contributors, can do to improve Homebrew apart from formulae upgrades? Like a list of ideas you’d like to see implemented in Homebrew but don’t have the time to?

**Comment by mikemcquaid** [comment]
@bfontaine Given that PR has stalled I think it'd be OK to reimplement. We'll see if they comment here though?

This is the closest thing to that list we have: https://github.com/Homebrew/homebrew/labels/help%20wanted (@xu-cheng may also be interested)

**Comment by mikemcquaid** [comment]
Going to pass on this given comments above. Sorry @bfontaine!

**Pull request closed without merging**